File: src/main/java/com/anod/appwatcher/installed/ImportBulkManager.java
package com.anod.appwatcher.installed;

import android.accounts.Account;
import android.content.Context;
import android.os.AsyncTask;
import android.support.v4.util.SimpleArrayMap;

import com.android.volley.VolleyError;
import com.anod.appwatcher.market.BulkDetailsEndpoint;
import com.anod.appwatcher.market.PlayStoreEndpoint;
import com.anod.appwatcher.model.AddWatchAppAsyncTask;
import com.anod.appwatcher.model.WatchAppList;
import com.google.android.finsky.api.model.Document;

import java.util.ArrayList;
import java.util.List;

class ImportBulkManager implements PlayStoreEndpoint.Listener, AddWatchAppAsyncTask.Listener {
    private static final int BULK_SIZE = 20;

    private final BulkDetailsEndpoint mEndpoint;
    private final WatchAppList mWatchAppList;
    private final Context mContext;
    private List<List<String>> listsDocIds;
    private int currentBulk;
    private ImportBulkManager.Listener mListener;
    private AsyncTask<Document, Void, SimpleArrayMap<String, Integer>> mTask;

    public interface Listener
    {
        void onImportProgress(List<String> docIds, SimpleArrayMap<String, Integer> result);
        void onImportFinish();
        void onImportStart(List<String> docIds);
    }

    ImportBulkManager(Context context, Listener listener) {
        mEndpoint = new BulkDetailsEndpoint(context);
        mEndpoint.setListener(this);
        mWatchAppList = new WatchAppList(null);
        mContext = context;
        mListener = listener;
    }

    public void init() {
        listsDocIds = new ArrayList<>();
        currentBulk = 0;
    }


    void stop() {
        mEndpoint.reset();
        if (mTask != null && !mTask.isCancelled()){
            mTask.cancel(true);
            mTask = null;
        }
    }

    void addPackage(String packageName) {
        List<String> currentList = null;
        if (listsDocIds.size() > currentBulk) {
            currentList = listsDocIds.get(currentBulk);
        }
        if (currentList == null) {
            currentList = new ArrayList<>();
            listsDocIds.add(currentList);
        } else if (currentList.size() > BULK_SIZE) {
            currentBulk++;
            currentList = new ArrayList<>();
            listsDocIds.add(currentList);
        }
        currentList.add(packageName);
    }

    public boolean isEmpty() {
        return listsDocIds == null || listsDocIds.size() == 0;
    }

    public void start() {
        currentBulk = 0;
        nextBulk();
    }

    private void nextBulk() {
        List<String> docIds = listsDocIds.get(currentBulk);
        mListener.onImportStart(docIds);
        mEndpoint.setDocIds(docIds);
        mEndpoint.startAsync();
    }

    @Override
    public void onDataChanged() {
        List<Document> docs = mEndpoint.getDocuments();
        mTask = new AddWatchAppAsyncTask(mWatchAppList, mContext, this).execute(docs.toArray(new Document[docs.size()]));
    }

    @Override
    public void onErrorResponse(VolleyError error) {
        List<String> docIds = listsDocIds.get(currentBulk);
        mListener.onImportProgress(docIds, new SimpleArrayMap<String, Integer>());
        currentBulk++;
        if (currentBulk == listsDocIds.size()) {
            mListener.onImportFinish();
        }
        else
        {
            nextBulk();
        }
    }

    void setAccount(Account account, String authSubToken) {
        mEndpoint.setAccount(account, authSubToken);
    }

    @Override
    public void onAddAppTaskFinish(SimpleArrayMap<String, Integer> result) {
        List<String> docIds = listsDocIds.get(currentBulk);
        mListener.onImportProgress(docIds, result);
        currentBulk++;
        if (currentBulk == listsDocIds.size()) {
            mListener.onImportFinish();
        }
        else
        {
            nextBulk();
        }
    }
}

File: app/src/main/java/com/anod/appwatcher/installed/ImportBulkManager.kt
// Copyright (c) 2019. Alex Gavrishev
package com.anod.appwatcher.installed

import androidx.collection.SimpleArrayMap
import finsky.api.BulkDocId
import info.anodsplace.applog.AppLog
import info.anodsplace.playstore.BulkDetailsEndpoint
import kotlinx.coroutines.Dispatchers.Main
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.withContext
import org.koin.core.Koin
import org.koin.core.parameter.parametersOf

sealed interface ImportStatus {
    object NotStarted : ImportStatus
    class Started(val docIds: List<String>) : ImportStatus
    class Progress(val docIds: List<String>, val result: SimpleArrayMap<String, Int>) : ImportStatus
    object Finished : ImportStatus
}

internal class ImportBulkManager(private val koin: Koin) {

    private var listsDocIds: MutableList<MutableList<BulkDocId>?> = mutableListOf()
    private var currentBulk: Int = 0
    private val importTask: ImportInstalledTask
        get() = koin.get()

    fun reset() {
        listsDocIds = mutableListOf()
        currentBulk = 0
    }

    fun addPackage(packageName: String, versionNumber: Int) {
        var currentList: MutableList<BulkDocId>? = null
        if (listsDocIds.size > currentBulk) {
            currentList = listsDocIds[currentBulk]
        }
        if (currentList == null) {
            currentList = ArrayList()
            listsDocIds.add(currentList)
        } else if (currentList.size >= BULK_SIZE) {
            currentBulk++
            currentList = ArrayList()
            listsDocIds.add(currentList)
        }
        currentList.add(BulkDocId(packageName, versionNumber))
    }

    val isEmpty: Boolean
        get() = listsDocIds.isEmpty()

    suspend fun start(): Flow<ImportStatus> = withContext(Main) {
        currentBulk = 0
        return@withContext flow {
            for (items in listsDocIds) {
                val docIds = items ?: continue
                if (docIds.isNotEmpty()) {
                    emit(ImportStatus.Started(docIds.map { it.packageName }))
                    val result = importDetails(docIds)
                    emit(ImportStatus.Progress(docIds.map { it.packageName }, result))
                }
            }
            emit(ImportStatus.Finished)
        }
    }

    private suspend fun importDetails(docIds: List<BulkDocId>): SimpleArrayMap<String, Int> = withContext(Main) {
        val endpoint = koin.get<BulkDetailsEndpoint> { parametersOf(docIds) }
        try {
            val model = endpoint.start()
            val docs = model.documents.toTypedArray()
            return@withContext importTask.execute(*docs)
        } catch (e: Exception) {
            AppLog.e(e)
            return@withContext SimpleArrayMap<String, Int>()
        }
    }

    companion object {
        private const val BULK_SIZE = 100
    }
}